sdk/typescript/src/transactions/plugins/NamedPackagesPlugin.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { SuiGraphQLClient } from '../../graphql/client.js';
+import type { NamedPackagesPluginCache, NameResolutionRequest } from '../../utils/move_registry.js';
+import {
+	findTransactionBlockNames,
+	listToRequests,
+	replaceNames,
+} from '../../utils/move_registry.js';
+import type { BuildTransactionOptions } from '../json-rpc-resolver.js';
+import type { TransactionDataBuilder } from '../TransactionData.js';
+
+export type NamedPackagesPlugin = {
+	/**
+	 * The SuiGraphQLClient to use for resolving names.
+	 * The endpoint should be the GraphQL endpoint of the network you are targeting.
+	 * For non-mainnet networks, if the plugin doesn't work as expected, you need to validate that the
+	 * RPC provider has support for the `packageByName` and `typeByName` queries (using external resolver).
+	 */
+	suiGraphQLClient: SuiGraphQLClient;
+	/**
+	 * The number of names to resolve in each batch request.
+	 * Needs to be calculated based on the GraphQL query limits.
+	 */
+	pageSize?: number;
+	/**
+	 * Local cache of the resolution plugin. Pass this to pre-populate
+	 * the cache with known packages / types (mainly useful for local or CI testing).
+	 *
+	 * Expected format is:
+	 * 1. For packages: `app@org` -> `0x1234`
+	 * 2. For types: `app@org::type::Type` -> `0x1234::type::Type`
+	 *
+	 */
+	cache?: NamedPackagesPluginCache;
```

Comment:
I'm a little conflicted about this.  I think I'd rather have this be `overrides`.

My thinking here is that cache implies that this object is cache implies that this is a cache instance that will get updated (which is how its currently implemented).  I don't think thats actually what I would want though.

eg if I had:

```ts
const LOCAL_RESOLUTIONS = {...}

tx.addPlugin(namedPackagesPlugin({ cache: LOCAL_RESOLUTIONS })
```

I am not sure I want those local definitions to be extended.  I can see the argument for wanting to have a cache instance you can share/control, but I think in that case, I'd want it to be a class or interface with methods.   If we want to have a plugable cache interface, we'd probably want to support something like a localStorage cache with support for async  APIs.

For now I think just having overrides, and then creating the cache internally is probably better, and we can look into adding a proper cache API in the future with methods to get/set/clear the cache

Reply:
> I am not sure I want those local definitions to be extended. 

To be honest, it seems reasonable to me to see it expand, and especially in this example of adding the plugin per transaction (vs global registration), it kinda feels better as you are really utilising previous transactions (versus starting clean with every transaction).

> we'd probably want to support something like a localStorage cache with support for async APIs.

I'd actually be careful on how we utilize persistent storage for caching packages (for types, it's fine), given that the non-versioned name is "latest', and "latest" can change. Also the network change results in different ids. I think the per-session time is fine to not move to the latest (and use the cache), but for a persistent storage we'd need to be super careful.

> then creating the cache internally is probably better

Assuming we indeed move with this approach (and the above arguments didn't sell 😆), is there any way to do this without introducing a "class" (which can create the cache obj on the constructor) for the inputs?